cpp/include/rapids_triton/triton/input.hpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <stdint.h>
+#include <triton/core/tritonbackend.h>
+#include <algorithm>
+#include <numeric>
+#include <rapids_triton/exceptions.hpp>
+#include <rapids_triton/tensor/dtype.hpp>
+#include <rapids_triton/utils/narrow.hpp>
+#include <sstream>
+#include <string>
+#include <vector>
+
+namespace triton {
+namespace backend {
+namespace rapids {
+inline auto* get_triton_input(TRITONBACKEND_Request* request, std::string const& name)
+{
+  auto result = static_cast<TRITONBACKEND_Input*>(nullptr);
+  triton_check(TRITONBACKEND_RequestInput(request, name.c_str(), &result));
+  return result;
+}
+
+template <typename T, typename Iter>
+auto get_triton_input_shape(Iter requests_begin, Iter requests_end, std::string const& name)
+{
+  auto result = std::vector<std::size_t>{};
+
+  auto reported_dtype     = DType{};
+  auto const* input_shape = static_cast<int64_t*>(nullptr);
+  auto input_dims         = uint32_t{};
+
+  auto batch_dim = std::accumulate(
+    requests_begin,
+    requests_end,
+    int64_t{},
+    [&reported_dtype, &input_shape, &input_dims, &name](auto total, auto& request) {
+      auto* input = get_triton_input(request, name);
+      triton_check(TRITONBACKEND_InputProperties(
+        input, nullptr, &reported_dtype, &input_shape, &input_dims, nullptr, nullptr));
+
+      if (reported_dtype != TritonDtype<T>::value) {
```

Comment:
`if constexpr`?

Reply:
I don't _think_ this can be `if constexpr` because `reported_dtype` can take on multiple runtime values. But maybe I'm thinking about that wrong?

Reply:
Yep, that causes a compiler error because `reported_dtype` is not a constant expression.